### PR TITLE
Upgrade reactive support for @NaturalId's and fix related tests

### PR DESCRIPTION
**Pull request by blafond**

### CODE REVIEW

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractPersisterDelegate.java [hunk 1]
```diff
@@ -45,21 +42,30 @@ public class ReactiveAbstractPersisterDelegate {
 	private final ReactiveSingleIdEntityLoader<Object> singleIdEntityLoader;
 	private final ReactiveMultiIdEntityLoader<?> multiIdEntityLoader;
 
+	private final ReactiveNaturalIdLoader<?> singleNaturalIdLoader;
+
+//	private final ReactiveMultiNaturalIdLoader<?> multiNaturalIdLoader;
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractPersisterDelegate.java:47]: Please, don't add comments like this. We haven't implemented this yet, just remove it

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveSingleTableEntityPersister.java [hunk 2]
```diff
@@ -104,6 +110,11 @@ public String[][] getLazyPropertyColumnAliases() {
 	public ReactiveSingleIdEntityLoader<?> getReactiveSingleIdEntityLoader() {
 		return reactiveDelegate.getSingleIdEntityLoader();
 	}
+//
+//	@Override
+//	public boolean didInitializeLazyProperty(String fieldName, Object entity, EntityEntry entry, int lazyIndex, Object selectedValue) {
+//		return ;
+//	}
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveSingleTableEntityPersister.java:117]: Please, remove this comment

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveCompoundNaturalIdLoader.java [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+            // TODO:  Uncomment and FIX
+//            applyNaturalIdRestriction(
+//                    bindValue,
+//                    rootTableGroup,
+//                    rootQuerySpec::applyPredicate,
+//                    jdbcParamBindings::addBinding,
+//                    sqlAstCreationState,
+//                    session
+//            );
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveCompoundNaturalIdLoader.java:219]: Why did you comment this?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveMultiNaturalIdLoaderStandard.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive.loader.ast.internal;
+
+import org.hibernate.loader.ast.internal.MultiNaturalIdLoaderStandard;
+import org.hibernate.metamodel.mapping.EntityMappingType;
+
+public class ReactiveMultiNaturalIdLoaderStandard<T> extends MultiNaturalIdLoaderStandard {
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveMultiNaturalIdLoaderStandard.java:11]: Are we using this class? I don't think we have implemented this

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveSimpleNaturalIdLoader.java [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+	public CompletionStage<Object> reactiveResolveIdToNaturalId(Object id, SharedSessionContractImplementor session) {
+		return delegate.reactiveResolveIdToNaturalId( id, session );
+	}
+
+	@Override
+	public CompletionStage<Object> reactiveResolveNaturalIdToId(
+			Object naturalIdValue,
+			SharedSessionContractImplementor session) {
+		// TODO: Move it to the delegator
+		return reactiveSelectByNaturalId(
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveSimpleNaturalIdLoader.java:102]: shouldn't you call the delegator here?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveSimpleNaturalIdLoader.java [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+		);
+	}
+
+	@Override
+	public CompletionStage<T> load(
+			Object naturalIdValue,
+			NaturalIdLoadOptions options,
+			SharedSessionContractImplementor session) {
+		// TODO: Move it to the delegator
+		return reactiveSelectByNaturalId(
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveSimpleNaturalIdLoader.java:139]: shouldn't you call the delegator here?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/spi/ReactiveMultiNaturalIdLoader.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    /**
+     * Load multiple entities by natural-id.  The exact result depends on the passed options.
+     *
+     * @param naturalIds The natural-ids to load.  The values of this array will depend on whether the
+     *                   natural-id is simple or complex.
+     * @param options
+     * @param session
+     */
+    @Override
+    <K> List<CompletionStage<T>> multiLoad(K[] naturalIds, MultiNaturalIdLoadOptions options, SharedSessionContractImplementor session);
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/spi/ReactiveMultiNaturalIdLoader.java:25]: I don't think we have implemented this yet, no need to add it to the PR

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java [hunk 8]
```diff
@@ -29,25 +16,28 @@
 import org.hibernate.collection.spi.PersistentCollection;
 import org.hibernate.dialect.Dialect;
 import org.hibernate.engine.internal.ManagedTypeHelper;
-import org.hibernate.engine.spi.EntityEntry;
-import org.hibernate.engine.spi.EntityKey;
-import org.hibernate.engine.spi.PersistenceContext;
-import org.hibernate.engine.spi.PersistentAttributeInterceptor;
-import org.hibernate.engine.spi.SessionFactoryImplementor;
-import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.engine.spi.*;
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:19]: We don't use `import *`

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java [hunk 9]
```diff
@@ -5,21 +5,8 @@
...
-import java.util.concurrent.CompletionStage;
-
-import org.hibernate.AssertionFailure;
-import org.hibernate.HibernateException;
-import org.hibernate.JDBCException;
-import org.hibernate.LockMode;
-import org.hibernate.LockOptions;
-import org.hibernate.StaleObjectStateException;
+import jakarta.persistence.metamodel.Attribute;
+import org.hibernate.*;
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:9]: we don't use `import *`

**Review by DavideD — CHANGES_REQUESTED** [review]
Please, don't add commented code.
1. I would not add the classes for the NaturalId multiLoad unless we have implemented it
2. Instead of new Unsupported, I've added the method `throw LOG.notYetImplemented()` . I think it's more explicit
3. Make sure to use the delegate in the natural id simple loader

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveCompoundNaturalIdLoader.java (hunk 3 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveCompoundNaturalIdLoader.java:219]: Having trouble uncommenting it and fixing the compile errors

rootQuerySpec::applyPredicate results in:  `Incompatible types: Object is not convertible to Predicate`
jdbcParamBindings::addBinding :  Incompatible types: Object is not convertible to JdbcParameter

Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ast/internal/ReactiveCompoundNaturalIdLoader.java:219]: Ok, I will have a look

**Comment by DavideD** [comment]
I'm going to send a new PR

**Comment by DavideD** [comment]
Superseeded by https://github.com/hibernate/hibernate-reactive/pull/1492

**Pull request closed without merging**